<a href="https://colab.research.google.com/github/jonaleal/ModelosRNA-Kmeans/blob/main/Modelo_RNA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import accuracy_score

In [ ]:
X, y = make_blobs(
        n_samples    = 500,
        n_features   = 2,
        centers      = 3,
        cluster_std  = 1.2,
        shuffle      = True,
        random_state = 0
       )

#Divición de datos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

# Crear un modelo de red neuronal
modelo = MLPClassifier(random_state=0,)
parameters = {'hidden_layer_sizes': [(5,), (10,), (15,), (5, 5), (10, 10), (15, 15), (5, 5, 5), (10, 10, 10), (15, 15, 15)],
              'activation': ['logistic'],
              'solver': ['sgd'],
              'max_iter': [2500]}

# Configurar GridSearchCV
grid_search = GridSearchCV(modelo, parameters, cv=5)

# Entrenar el modelo
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=MLPClassifier(random_state=0),
             param_grid={'activation': ['logistic'],
                         'hidden_layer_sizes': [(5,), (10,), (15,), (5, 5),
                                                (10, 10), (15, 15), (5, 5, 5),
                                                (10, 10, 10), (15, 15, 15)],
                         'max_iter': [2500], 'solver': ['sgd']})

In [ ]:
# Obtener los resultados de la búsqueda de GridSearch
results = grid_search.cv_results_
params = results['params']
mean_test_score = results['mean_test_score']

#Extraer la cantidad de neuronas de las configuraciones
neuron_counts = [param['hidden_layer_sizes'] if isinstance(param['hidden_layer_sizes'], tuple) else (param['hidden_layer_sizes'],) for param in params]

#Imprimir la precisión promedio para cada configuración de neuronas
print("Precisión promedio para cada configuración de neuronas:")
for neurons, score in zip(neuron_counts, mean_test_score):
    print(f"Neuronas: {neurons}, Precisión: {score:.4f}")


Precisión promedio para cada configuración de neuronas:
Neuronas: (5,), Precisión: 0.860
Neuronas: (10,), Precisión: 0.874
Neuronas: (15,), Precisión: 0.877
Neuronas: (5, 5), Precisión: 0.217
Neuronas: (10, 10), Precisión: 0.174
Neuronas: (15, 15), Precisión: 0.346
Neuronas: (5, 5, 5), Precisión: 0.366
Neuronas: (10, 10, 10), Precisión: 0.346
Neuronas: (15, 15, 15), Precisión: 0.346


In [ ]:
print("Mejor combinación de hiperparámetros:", grid_search.best_params_)
print("Mejor puntuación de validación cruzada:", grid_search.best_score_)

Mejor combinación de hiperparámetros: {'activation': 'logistic', 'hidden_layer_sizes': (15,), 'max_iter': 2500, 'solver': 'sgd'}
Mejor puntuación de validación cruzada: 0.8771428571428572


Aunque el mejor Modelo para Grid Search sea el que tiene 1 capa y 15 neuronas, Para nosotros el mejor modelo sería el de 1 capa y 10 neuronas puesto que es más simple pero entrega un resultado con muy poca diferentecia de precision tan solo 0.003.

In [ ]:
#Volver a Entrenar con el modelo elegido
hyperparameters = {'activation': 'logistic', 'hidden_layer_sizes': (10,), 'max_iter': 2500, 'solver': 'sgd'}

# Crear un nuevo modelo con los hiperparámetros especificados
specific_model = MLPClassifier(**hyperparameters)

# Entrenar el modelo con tus datos de entrenamiento
specific_model.fit(X_train, y_train)

# Hacer predicciones con el modelo
predictions_specific = specific_model.predict(X_test)

# Calcular la precisión del modelo
accuracy_specific = accuracy_score(y_test, predictions_specific)
print(f"Precisión del modelo específico: {accuracy_specific:.4f}")

Precisión del modelo específico: 0.8667


Podemos ver que la Precisión del modelo a la hora de testear es 0.8667 mientras que en el entrenamiento fue 0.874, la diferencia es muy poco y en este caso es menor a la de entrenamiento lo que significa que el modelo no esta sobreajustado